mailpoet/assets/js/src/segments/dynamic/subscribers_counter.tsx

```diff
@@ -24,34 +24,62 @@ const SubscribersCounter: React.FunctionComponent = () => {
   const { updateSubscriberCount } = useDispatch('mailpoet-dynamic-segments-form');
 
   const serializedSegment = JSON.stringify(segment);
+  const latestRequestIdRef = useRef(1);
+  const deferredRequestRef = useRef(null);
+  const isRequestInFlight = useRef(false);
+
+  function load(loadItem: Segment): void {
+    // Don't allow multiple in-flight requests to avoid hammering the database
+    // when we'll only ever use the results of the last request
+    if (isRequestInFlight.current) {
+      deferredRequestRef.current = loadItem;
+      return;
+    }
+
+    deferredRequestRef.current = null;
+    latestRequestIdRef.current += 1;
+    const requestId = latestRequestIdRef.current;
+    isRequestInFlight.current = true;
+
+    loadCount(loadItem).then((response) => {
+      isRequestInFlight.current = false;
+      if (deferredRequestRef.current) {
+        load(deferredRequestRef.current);
+        return;
+      }
+      if (requestId !== latestRequestIdRef.current) {
+        // Don't do anything with the response because a newer request has been initiated
+        return;
+      }
```

Comment:
I don't have the context of what happens in the UI when the data is updated but maybe we could do `if (requestId < latestRequestIdRef.current)` here to avoid outdated requests doing anything but display some changes for non-outdated ones so at least some changes are displayed? Not sure, maybe not too important.

Reply:
I'm not sure I completely follow. How would we get into a situation where there would be a difference between `requestId < latestRequestIdRef.current` and `requestId !== latestRequestIdRef.current`?

In any case, because this `return`s before calling `updateSubscriberCount`, the UI should continue to display the `Calculating segment size…` placeholder text until the latest request finishes, which makes sense to me.

I also [moved the code](http://github.com/mailpoet/mailpoet/blob/update/MAILPOET-4018/debounce_dynamic_segments/mailpoet/assets/js/src/segments/dynamic/subscribers_counter.tsx#L76-L76) that starts displaying that placeholder text outside of `load`, because I wanted to avoid confusing situations like this:

1. User makes change at t=0 that will complete at t=3 (2 seconds delay + 1 second for the Ajax request)
2. User makes another change at t=2.5, which won't trigger the debounced `load` until t=4.5
3. At t=3, the first request comes back and updates the count. It would look like this number represents the count for the configuration on the screen
4. At t=4.5, 1.5 seconds later, the second debounced `load` function fires, and all of a sudden the form is recalculating without the user having made a change.

I could have moved out the incrementing of `latestRequestIdRef.current` instead  and passed the request ID as a parameter into `load`, but I thought it would be a better user experience to start showing "Calculating segment size…" as soon as a user does something that we know 100% will result in us eventually calculating the segment size. It's immediate feedback even though we know nothing will happen for at least 2 seconds. The only downside I could see is that it could make it look like all of our calculations are really slow.